File: year2024/day02/typescript/solution.ts
import { readFileSync } from 'fs';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const input: number[][] = file.trim().split('\n').map(l => l.split(' ').map(Number));
if(process.argv[2] === 'puzzle') {
  console.log(part_1());
  console.log(part_2());
}

export function part_1(): number {
  return input.filter(isSafe).length;
}

export function part_2(): number {
  return input.filter(r => isSafe(r) || r.some((_, i) => isSafe(r.filter((_, j) => j !== i)))).length;
}

function isSafe(levels: number[]): boolean {
  const diffs = levels.slice(1).map((n, i) => n - levels[i]);
  return (
    diffs.every(d => Math.abs(d) >= 1 && Math.abs(d) <= 3) &&
    (diffs.every(d => d > 0) || diffs.every(d => d < 0))
  );
}

File: year2024/day02/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
    return len([levels for levels in input if is_safe(levels)])

def part_2():
    return len([
        levels for levels in input
        if is_safe(levels) or any(
            is_safe([levels[j] for j in range(len(levels)) if j != i])
            for i in range(len(levels))
        )
    ])

def is_safe(levels):
    diffs = [levels[i + 1] - levels[i] for i in range(len(levels) - 1)]
    return (
        all(1 <= abs(d) <= 3 for d in diffs) and
        (all(d > 0 for d in diffs) or all(d < 0 for d in diffs))
    )



def setup(path):
  global input

  with open(path, 'r') as file:
    input = [[int(item) for item in line.split()] for line in file]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()